lib/Sema/TypeCheckCodeCompletion.cpp

```diff
@@ -592,20 +592,282 @@ TypeChecker::getTypeOfCompletionOperator(DeclContext *DC, Expr *LHS,
+      }
+
+      if (isa<ApplyExpr>(E)) {
+        Contexts.push_back(std::make_pair(ContextKind::Application, E));
+      }
+
+      if (isa<CodeCompletionExpr>(E)) {
+        CompletionExpr = E;
+        return std::make_pair(false, nullptr);
+      }
+
+      if (auto *Error = dyn_cast<ErrorExpr>(E)) {
+        Contexts.push_back(std::make_pair(ContextKind::ErrorExpression, E));
+        if (auto *OrigExpr = Error->getOriginalExpr()) {
+          OrigExpr->walk(*this);
+          return std::make_pair(false, hasCompletionExpr() ? nullptr : E);
+        }
+      }
+
+      return std::make_pair(true, E);
+    }
+
+    Expr *walkToExprPost(Expr *E) override {
+      if (isa<ClosureExpr>(E) || isa<InterpolatedStringLiteralExpr>(E) ||
+          isa<ApplyExpr>(E) || isa<ErrorExpr>(E))
+        Contexts.pop_back();
+      return E;
+    }
+
+    /// Check whether code completion expression is located inside of a
+    /// multi-statement closure.
+    bool locatedInMultiStmtClosure() const {
+      return hasContext(ContextKind::MultiStmtClosure);
+    }
+
+    bool locatedInStringIterpolation() const {
+      return hasContext(ContextKind::StringInterpolation);
+    }
+
+    bool hasCompletionExpr() const {
+      return CompletionExpr;
+    }
+
+    Expr *getCompletionExpr() const {
+      assert(CompletionExpr);
+      return CompletionExpr;
+    }
+
+    ErrorExpr *getInnermostErrorExpr() const {
+      for (const Context &curr : llvm::reverse(Contexts)) {
+        if (curr.first == ContextKind::ErrorExpression)
+          return cast<ErrorExpr>(curr.second);
+      }
+      return nullptr;
+    }
+
+    ClosureExpr *getOutermostMultiStmtClosure() const {
+      for (const Context &curr : Contexts) {
+        if (curr.first == ContextKind::MultiStmtClosure)
+          return cast<ClosureExpr>(curr.second);
+      }
+      return nullptr;
+    }
+
+    /// As a fallback sometimes its useful to not only type-check
+    /// code completion expression directly but instead add some
+    /// of the enclosing context e.g. when completion is an argument
+    /// to a call.
+    Expr *getCompletionExprInContext() const {
+      assert(CompletionExpr);
+
+      auto &innerContext = Contexts.back();
+      return innerContext.first == ContextKind::Application
+                 ? innerContext.second
+                 : CompletionExpr;
+    }
+
+  private:
+    bool hasContext(ContextKind kind) const {
+      return llvm::find_if(Contexts, [&kind](const Context &currContext) {
+               return currContext.first == kind;
+             }) != Contexts.end();
+    }
+  };
+
+  ContextFinder contextAnalyzer(expr);
+  expr->walk(contextAnalyzer);
+
+  // If there was no completion expr (e.g. if the code completion location is
+  // within an ErrorExpr without an valid subexpr due to parser error recovery)
+  // bail.
+  if (!contextAnalyzer.hasCompletionExpr())
+    return false;
```

Comment:
The original expression in `ErrorExpr` is now walked into. If you know other cases, please update the comment.

Reply:
Right, I think this branch is impossible to get into at the moment since we're only doing this for member completion. Do you know if once we change the other completion kinds to use this approach there will ever be a case where no CodeCompletionExpr is formed within an ErrorExpr? I assumed this was possible, e.g. if the completion location is in the middle of some nonsensical tokens between statement starting tokens, but if we're always guaranteed to have a CodeCompletionExpr regardless of completion kind I think this should be an assertion.

Reply:
I don’t know off the top of my head. But I think we should keep this `return false` path, just to be on the safe side.